# DEVEA
an interactive shiny application for **D**ifferential **E**xpression analysis, data **V**isualization and **E**nrichment **A**nalysis of **transcriptomics data**.

We are at a time of considerable growth in transcriptomics studies and subsequent in silico analysis. RNA sequencing (RNA-Seq) is the most widely used approach to analyse the transcriptome and is integrated in many studies. 

The processing of transcriptomic data typically requires a noteworthy number of steps, statistical knowledge, and coding skills, which are not accessible to all scientists. Despite the development of a plethora of software applications over the past few years to address this concern, there is still room for improvement. 

Here we present DEVEA, an R shiny application tool developed to perform differential expression analysis, data visualization and enrichment pathway analysis mainly from transcriptomics data, but also from simpler gene lists with or without statistical values. 

The intuitive and easy-to-manipulate interface facilitates gene expression exploration through numerous interactive figures and tables, and statistical comparisons of expression profile levels between groups. Further meta-analysis such as enrichment analysis is also possible, without the need for prior bioinformatics expertise.

DEVEA performs a comprehensive analysis from multiple and flexible data sources representing distinct analytical steps. Consequently, it produces dynamic graphs and tables, to explore the expression levels and statistical results from differential expression analysis. Moreover, it generates a comprehensive pathway analysis to extend biological insights. Finally, a complete and customizable HTML report can be extracted to enable the scientists to explore results beyond the application. DEVEA is freely accessible at **https://shiny.imib.es/devea/**. 

For some indications on how to download and launch the tool locally, please refer to the **Supplemental-Information.md** file. Notice that some 'Demo files' for several species to test the app are available in the **\data** folder on this repository. If you want to generate your own data, you may check **Galaxy_tutorials.md** containing key information on how to build your dataset from raw data.

Finaly, a detailed walk-through with visual informacion regarding how to use DEVEA is available on **tutorial.html**.


For more information, you can check: Riquelme-Perez M, Perez-Sanz F, Deleuze JF et al. DEVEA: an interactive shiny application for Differential Expression analysis, data Visualization and Enrichment Analysis of transcriptomics data [version 1; peer review: 1 approved, 1 approved with reservations]. F1000Research 2022, 11:711 (https://doi.org/10.12688/f1000research.122949.1)
 
 [![DOI](https://zenodo.org/badge/501792238.svg)](https://zenodo.org/badge/latestdoi/501792238)

